aesd-char-driver/main.c

```diff
@@ -41,35 +52,172 @@ int aesd_release(struct inode *inode, struct file *filp)
 	/**
 	 * TODO: handle release
 	 */
+	 
 	return 0;
 }
 
-ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
-                loff_t *f_pos)
+ssize_t aesd_read(struct file *filp, char __user *buf, size_t count, loff_t *f_pos)
 {
 	ssize_t retval = 0;
-	PDEBUG("read %zu bytes with offset %lld",count,*f_pos);
-	/**
-	 * TODO: handle read
-	 */
+	
+	struct aesd_dev *l_dev = NULL;
+	struct aesd_buffer_entry *loc_hit = NULL;
+  size_t off_byte_pos = 0;
+  size_t bytes_to_be_read_out = 0;
+  size_t bytes_failed_to_copy = 0;
+	
+  PDEBUG("read %zu bytes with offset %lld",count,*f_pos);
+	//fetch the content received from the script
+	l_dev = (struct aesd_dev *)filp->private_data;
+	
+	//take mutex
+	if(mutex_lock_interruptible(&(l_dev->device_lock)))
```

Comment:
@Rishab-Shah you don't need the extra parens here on the lock reference, can use `&l_dev->device_lock` instead.